src/Serialization.SystemTextJson/GraphQLRequestDeserializer.cs

```diff
@@ -0,0 +1,92 @@
+using GraphQL.Server.Common;
+using System;
+using System.Buffers;
+using System.IO;
+using System.Text.Json;
+using System.Threading.Tasks;
+
+namespace GraphQL.Server.Serialization.SystemTextJson
+{
+    public class GraphQLRequestDeserializer : IGraphQLRequestDeserializer
+    {
+        /// <summary>
+        /// Max size of a buffer we can create using the Shared ArrayPool.
+        /// </summary>
+        /// <remarks>
+        /// Determined by following <see cref="ArrayPool{T}.Shared" /> (https://github.com/dotnet/coreclr/blob/master/src/System.Private.CoreLib/shared/System/Buffers/ArrayPool.cs)
+        /// which instantiates a TlsOverPerCoreLockedStacksArrayPool (https://github.com/dotnet/runtime/blob/ccf6aedb63c37ea8e10e4f5b5d9d23a69bdd9489/src/libraries/System.Private.CoreLib/src/System/Buffers/TlsOverPerCoreLockedStacksArrayPool.cs)
+        /// whose ctor calls Utilities.GetMaxBucketSize with TlsOverPerCoreLockedStacksArrayPool.NumBuckets = 17 (https://github.com/dotnet/runtime/blob/ccf6aedb63c37ea8e10e4f5b5d9d23a69bdd9489/src/libraries/System.Private.CoreLib/src/System/Buffers/Utilities.cs#L22)
+        /// which does 16 << 17 = 2MB as below.
+        /// </remarks>
+        private const int MaxSharedBufferSize = 2097152; 
```

Comment:
I believe that using this constant in our case simply does not make sense, you should always go to the pool for an array and that’s it. It only complicates the code.

Reply:
I think David's point here was that if you go over this it'll be heap allocated (which I assume is slower?) though maybe that's still better than falling back into the MemoryStream branch. Sorry, this kind of stuff really isn't my forte.

I can't imagine a graphql request over 2MB is a thing; but maybe it is if people do file uploads by sending base64 encoded image content in variables.

Reply:
You don't want large single contiguous arrays allocated per request (they'll each end up on the large object heap which has gen 2 clean up semantics). This entire code path trying to optimize smaller payloads should be measured though...

Reply:
> I think David's point here was that if you go over this it'll be heap allocated (which I assume is slower?)

Yes, heap allocated. And what is the alternative? In **both** cases we have heap allocation :) The only difference is with what API we come to the call `new` operator. Why explicitly do another branch of code? The difference in performance is within the margin of error.

https://adamsitnik.com/Array-Pool/#bufferallocated

Reply:
> You don't want large single contiguous arrays allocated per request (they'll each end up on the large object heap which has gen 2 clean up semantics). This entire code path trying to optimize smaller payloads should be measured though...

Here we smoothly move to another optimization plane (Pipes) and go beyond the scope of this PR (async deserialization). Let's finish this PR first and not mix concepts.

Reply:
One heap allocation is pooled and reused, the other sits in memory for way longer than the request lasts and isn’t reused for a much longer time (which is the inherent problem). 

None of these should be concerns for System m.Text.Json because it’s fully async.

Reply:
I only say that there is no fundamental difference compared to a code branch that copies bytes to a stream and then obtains reference to byte[] buffer. I do not argue that switching to pipes will be better.

Reply:
As soon as this PR is merged we'll be looking to use it in production so we can get onto 3.1.x, so I'm keen to not only address the async issue but also hit highest perf in one go. 

I'm more than happy to have a crack at doing it with Pipelines. That would (I think):
* Let us use one way to handle buffering, avoiding the if-else and if there's a way to set the buffer size to content length then perfect.
* Let us, in the while loop, check the first byte if it's `{` or `[` and bail out or then give us the target type to deserialize into
* Finally, with the target type, and after buffering the whole request, then we can deserialize using the Deserialize overload that takes `ReadOnlySpan<byte> utf8Json`.

Reply:
Ah, just saw the Peek alternative above, in which case, we can use Deserialize on the stream direct and all these problems go away. Brilliant!

Reply:
Yes, I'm happy too.